dimod/binary/pybqm.py

```diff
@@ -0,0 +1,324 @@
+                    self.add_quadratic(u, v, quadratic[u, v])
+
+    def add_variable(self, v: Optional[Variable] = None,
+                     bias: Any = 0) -> Variable:
+        if v is None:
+            # match the behaviour of dimod.Variables
+            adj = self._adj
+
+            v = len(adj)
+
+            if v in adj:
+                v = 0
+                while v in adj:
+                    v += 1
+
+        self.add_linear(v, bias)
+        return v
+
+    def change_vartype(self, vartype: VartypeLike) -> 'pyBQM':
+        vartype = as_vartype(vartype)
+
+        # in place and we are already correct, so nothing to do
+        if self.vartype == vartype:
+            return self
+
+        if vartype == Vartype.BINARY:
+            lin_mp = 2.
+            lin_offset_mp = -1.
+            quad_mp = 4.
+            lin_quad_mp = -2.
+            quad_offset_mp = .5
+        elif vartype == Vartype.SPIN:
+            lin_mp = .5
+            lin_offset_mp = .5
+            quad_mp = .25
+            lin_quad_mp = .25
+            quad_offset_mp = .125
+        else:
+            raise RuntimeError("unexpected vartype")
+
+        adj = self._adj
+
+        for u, Nu in adj.items():
+            lbias = Nu[u]
+
+            self.offset += lin_offset_mp * lbias
+            Nu[u] = lin_mp * lbias
+
+            for v, qbias in Nu.items():
+                if v == u:
+                    continue
+                Nu[v] = quad_mp * qbias
+                Nu[u] += lin_quad_mp * qbias  # linear
+                self.offset += quad_offset_mp * qbias
+
+        self._vartype = vartype
+
+        return self
+
+    def degree(self, v: Variable) -> int:
+        try:
+            return len(self._adj[v]) - 1
+        except KeyError:
+            raise ValueError(f"unknown variable {v!r}") from None
+
+    def energies(self, samples_like, dtype: DTypeLike = None):
+        samples, labels = as_samples(samples_like)
+
+        if len(labels) != self.num_variables():
+            raise ValueError("variable_order does not match the number of "
+                             "variables")
+
+        ldata = np.asarray([self.get_linear(v) for v in labels])
+
+        label_to_idx = {v: idx for idx, v in enumerate(labels)}
+        irow = []
+        icol = []
+        qdata = []
+        for u, v, bias in self.iter_quadratic():
+            irow.append(label_to_idx[u])
+            icol.append(label_to_idx[v])
+            qdata.append(bias)
+
+        irow = np.asarray(irow)  # type: ignore[assignment]
+        icol = np.asarray(icol)  # type: ignore[assignment]
+        qdata = np.asarray(qdata)  # type: ignore[assignment]
+
+        energies = samples.dot(ldata)
+        energies += (samples[:, irow]*samples[:, icol]).dot(qdata)
+        energies += ldata.dtype.type(self.offset)
+
+        return np.asarray(energies, dtype=dtype)
+
+    def get_linear(self, v: Variable) -> Any:
+        try:
+            return self._adj[v][v]
```

Comment:
Separating out the linear from _adj has its own advantages during simulated annealing. But I guess this bqm is not going to be used for simulated annealing.

Reply:
yeah, in this case this is just meant to be a simple python BQM, not really useful for solving on.